### PR TITLE
Add the synchronous gauge to the metric API and SDK

### LINKED ISSUE #5225: Support synchronous gauge
### Problem Statement

From [specification](https://opentelemetry.io/docs/specs/otel/metrics/api/#gauge), `gauge` can be synchronous or asynchronous.

> Gauge is a [synchronous Instrument](https://opentelemetry.io/docs/specs/otel/metrics/api/#synchronous-instrument-api) which can be used to record non-additive value(s) (e.g. the background noise level - it makes no sense to record the background noise level value from multiple rooms and sum them up) when changes occur.

### Proposed Solution

Can we add support for sync gauge?

### PR DESCRIPTION
**Pull request by MrAlias**
Resolve #5225 

The specification has [added a synchronous gauge instrument](https://github.com/open-telemetry/opentelemetry-specification/pull/3540). That instrument has now been [stabilized](https://github.com/open-telemetry/opentelemetry-specification/pull/4019), and that stabilization is included in the [next release](https://github.com/open-telemetry/opentelemetry-specification/pull/4034).

This adds the new synchronous gauge instrument to the metric API and all implementation we publish.

This change will be a breaking change for any SDK developer. The `embedded` package is updated to ensure our compatibility guarantees are meet.

### CODE REVIEW

File: sdk/metric/pipeline_registry_test.go [hunk 1]
```diff
@@ -699,6 +718,32 @@ func TestIsAggregatorCompatible(t *testing.T) {
 			kind: InstrumentKindHistogram,
 			agg:  AggregationBase2ExponentialHistogram{},
 		},
+		{
+			name: "SyncGauge and Drop",
+			kind: InstrumentKindGauge,
+			agg:  AggregationDrop{},
+		},
+		{
+			name: "SyncGauge and LastValue{}",
```
Comment (MrAlias) [inline sdk/metric/pipeline_registry_test.go:727]: Don't copy naming cruft

```suggestion
			name: "SyncGauge and LastValue",
```

**Comment by dashpole** [comment]
Does this handle the Delta vs Cumulative temporality distinction for gauges?  (from https://github.com/open-telemetry/opentelemetry-specification/pull/3540)

> The MetricReader selection of temporality as a function of instrument kind influences the persistence of metric data points across collections. For synchronous instruments with Cumulative aggregation temporality, [MetricReader.Collect](https://github.com/open-telemetry/opentelemetry-specification/blob/6360b49d20ae451b28f7ba0be168ed9a799ac9e1/specification/metrics/sdk.md#collect) MUST receive data points exposed in previous collections regardless of whether new measurements have been recorded. For synchronous instruments with Delta aggregation temporality, [MetricReader.Collect](https://github.com/open-telemetry/opentelemetry-specification/blob/6360b49d20ae451b28f7ba0be168ed9a799ac9e1/specification/metrics/sdk.md#collect) MUST only receive data points with measurements recorded since the previous collection. For asynchronous instruments with Delta or Cumulative aggregation temporality, [MetricReader.Collect](https://github.com/open-telemetry/opentelemetry-specification/blob/6360b49d20ae451b28f7ba0be168ed9a799ac9e1/specification/metrics/sdk.md#collect) MUST only receive data points with measurements recorded since the previous collection. These rules apply to all metrics, not just those whose [point kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/6360b49d20ae451b28f7ba0be168ed9a799ac9e1/specification/metrics/data-model.md#point-kinds) includes an aggregation temporality field.

**Comment by MrAlias** [comment]
> Does this handle the Delta vs Cumulative temporality distinction for gauges? (from [open-telemetry/opentelemetry-specification#3540](https://github.com/open-telemetry/opentelemetry-specification/pull/3540))

It does not.

I'm going to submit another PR to address the `sdk/metric/internal/aggregate` package. That PR will address that and the timestamps for last-values.

Moving this to a draft until that is merged.

**Comment by MrAlias** [comment]
> I'm going to submit another PR to address the `sdk/metric/internal/aggregate` package. That PR will address that and the timestamps for last-values.

https://github.com/open-telemetry/opentelemetry-go/pull/5305

File: metric/noop/noop.go [hunk 2]
```diff
@@ -112,6 +120,12 @@ func (Meter) Float64Histogram(string, ...metric.Float64HistogramOption) (metric.
 	return Float64Histogram{}, nil
 }
 
+// Float64Gauge returns a Gauge used to record int64 measurements that
```
Comment (dashpole) [inline metric/noop/noop.go:123]:
```suggestion
// Float64Gauge returns a Gauge used to record float64 measurements that
```

**Review by dashpole — APPROVED** [review]

**Review by MadVikingGod — APPROVED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -14,6 +14,9 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Add `RecordFactory` in `go.opentelemetry.io/otel/sdk/log/logtest` to facilitate testing the exporter and processor implementations. (#5258)
 - Add example for `go.opentelemetry.io/otel/exporters/stdout/stdoutlog`. (#5242)
 - The count of dropped records from the `BatchProcessor` in `go.opentelemetry.io/otel/sdk/log` is logged. (#5276)
+- Add the synchronous gauge instrument to `go.opentelemetry.io/otel/metric`. (#5304)
```
Comment (pellared) [inline CHANGELOG.md:17]: (non-blocking)

What do you think of calling out all the added exported API? I mean do you think that it is worth mentioning that this PR:
- adds `Int64Gauge` and `Float64Gauge` methods to `Meter` interface.
- adds `Int64Gauge` and `Float64Gauge` interfaces
- adds `*GaugeConfig` structs, factory functions, options...

**Review by pellared — APPROVED** [review]

**Pull request merged**